File: test/health.test.mjs
import { describe, it } from "node:test";
import assert from "node:assert/strict";
import { parseCacheFixHealth } from "../src/health.mjs";

describe("parseCacheFixHealth", () => {
  it("ok with version", () => {
    const r = parseCacheFixHealth(
      200,
      JSON.stringify({ status: "ok", version: "4.3.0", forward_proxy: false }),
    );
    assert.equal(r.kind, "ok");
    assert.equal(r.version, "4.3.0");
    assert.equal(r.forwardProxy, false);
  });

  it("degraded with 503", () => {
    const r = parseCacheFixHealth(
      503,
      JSON.stringify({
        status: "degraded",
        version: "4.3.1",
        failed_extensions: ["x"],
        hint: "check logs",
      }),
    );
    assert.equal(r.kind, "degraded");
    assert.equal(r.httpStatus, 503);
    assert.deepEqual(r.failed_extensions, ["x"]);
    assert.equal(r.hint, "check logs");
  });

  it("ok with only forward_proxy (no version)", () => {
    const r = parseCacheFixHealth(
      200,
      JSON.stringify({ status: "ok", forward_proxy: true }),
    );
    assert.equal(r.kind, "ok");
    assert.equal(r.forwardProxy, true);
  });

  it("foreign for unknown JSON / missing recognition fields", () => {
    assert.equal(
      parseCacheFixHealth(200, JSON.stringify({ status: "ok" })).kind,
      "foreign",
    );
    assert.equal(
      parseCacheFixHealth(200, JSON.stringify({ healthy: true })).kind,
      "foreign",
    );
    assert.equal(
      parseCacheFixHealth(200, JSON.stringify([1, 2, 3])).kind,
      "foreign",
    );
    assert.equal(parseCacheFixHealth(200, "not-json").kind, "foreign");
  });

  it("unreachable when no status/body", () => {
    assert.equal(parseCacheFixHealth(null, null).kind, "unreachable");
    assert.equal(parseCacheFixHealth(undefined, "").kind, "unreachable");
  });
});

File: shared/src/commonTest/kotlin/com/cachefix/gui/HealthTest.kt
package com.cachefix.gui

import com.cachefix.gui.health.HealthKind
import com.cachefix.gui.health.parseCacheFixHealth
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNull

class HealthTest {
    @Test
    fun okWithVersion() {
        val r = parseCacheFixHealth(
            200,
            """{"status":"ok","version":"4.3.0","forward_proxy":false}""",
        )
        assertEquals(HealthKind.OK, r.kind)
        assertEquals("4.3.0", r.version)
        assertEquals(false, r.forwardProxy)
    }

    @Test
    fun degradedWith503() {
        val r = parseCacheFixHealth(
            503,
            """{"status":"degraded","version":"4.3.1","failed_extensions":["x"],"hint":"check logs"}""",
        )
        assertEquals(HealthKind.DEGRADED, r.kind)
        assertEquals(503, r.httpStatus)
        assertEquals("check logs", r.hint)
        @Suppress("UNCHECKED_CAST")
        assertEquals(listOf("x"), r.failedExtensions as List<String>)
    }

    @Test
    fun okWithOnlyForwardProxy() {
        val r = parseCacheFixHealth(
            200,
            """{"status":"ok","forward_proxy":true}""",
        )
        assertEquals(HealthKind.OK, r.kind)
        assertEquals(true, r.forwardProxy)
        assertNull(r.version)
    }

    @Test
    fun foreignForUnknownJson() {
        assertEquals(
            HealthKind.FOREIGN,
            parseCacheFixHealth(200, """{"status":"ok"}""").kind,
        )
        assertEquals(
            HealthKind.FOREIGN,
            parseCacheFixHealth(200, """{"healthy":true}""").kind,
        )
        assertEquals(
            HealthKind.FOREIGN,
            parseCacheFixHealth(200, """[1,2,3]""").kind,
        )
        assertEquals(HealthKind.FOREIGN, parseCacheFixHealth(200, "not-json").kind)
    }

    @Test
    fun unreachableWhenNoStatusOrBody() {
        assertEquals(HealthKind.UNREACHABLE, parseCacheFixHealth(null, null).kind)
        assertEquals(HealthKind.UNREACHABLE, parseCacheFixHealth(null, "").kind)
    }
}
